syntax = "proto3";

package books_database;

message BookInfo {
  string title = 1;
  int32 stock = 2;
  int32 version = 3;
}

message ReadRequest {
  string title = 1;
}

message WriteRequest {
  string title = 1;
  int32 stock = 2;
  int32 version = 3;
}

message WriteResponse {
  bool success = 1;
}

message OpenTransactionResponse {
  string transaction_id = 1;
}

message OpenTransactionRequest {}

message InitTwoPhaseCommitResponse {
  string status = 1;
}

message InitTwoPhaseCommitRequest {
  string transaction_id = 1;
}

message AbortTwoPhaseCommitResponse {
  string status = 1;
}

message AbortTwoPhaseCommitRequest {
  string transaction_id = 1;
}

message CommitTwoPhaseCommitResponse {
  string status = 1;
}

message CommitTwoPhaseCommitRequest {
  string transaction_id = 1;
}

service BooksDatabaseService {
  rpc Read (ReadRequest) returns (BookInfo) {}
  rpc Write (WriteRequest) returns (WriteResponse) {}
  rpc SyncWrite (WriteRequest) returns (WriteResponse) {}
  rpc OpenTransaction (OpenTransactionRequest) returns (OpenTransactionResponse) {}
  rpc InitTwoPhaseCommit (InitTwoPhaseCommitRequest) returns (InitTwoPhaseCommitResponse) {}
  rpc AbortTwoPhaseCommit (AbortTwoPhaseCommitRequest) returns (AbortTwoPhaseCommitResponse) {}
  rpc CommitTwoPhaseCommit (CommitTwoPhaseCommitRequest) returns (CommitTwoPhaseCommitResponse) {}
}